### PR TITLE
[DDC-2363] Duplicated record with orphanRemoval and proxy

### PR DESCRIPTION
**Pull request by mmenozzi**
See http://www.doctrine-project.org/jira/browse/DDC-2363.

### CODE REVIEW

**Comment by Ocramius** [comment]
Note: retried this on `2.6.0` and it still fails.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        // To make test pass, uncomment the following line. So ServicesPackage will be hydrated.
+        // $order->getServicesPackage()->getName();
+
+        // ... and create another ServicesPackage with two Services.
+        $servicesPackage = $this->createNewServicesPackage($this->_em);
+        $order->setServicesPackage($servicesPackage);
+
+        $this->_em->persist($order);
+        $this->_em->flush();
+        $this->_em->flush(); // <--- This is the problem!
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: Something else is happening here. I went through it with the debugger, and what is happening is that the `$services_package` property is replaced by the one created at [line 60](https://github.com/doctrine/doctrine2/pull/625/files#diff-d96f5038a94052e8ae58f5a7c642fa3cR60), so this second `flush()` is working with an incorrect object graph.

The [first `flush()`](https://github.com/doctrine/doctrine2/pull/625/files#diff-d96f5038a94052e8ae58f5a7c642fa3cR93) is causing the graph to be broken

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: The problem seems to be that the first `flush()` causes lazy-loading of the proxy previously referenced in the `services_package` property. This is correct, as these entities are targeted for removal. The lazy-loading is forced by `cascadeRemove()`:

https://github.com/doctrine/doctrine2/blob/374e7ace49d864dad8cddbc55346447c8a6a2083/lib/Doctrine/ORM/UnitOfWork.php#L2300-L2313

Lazy-loading of a `DDC2363ServicesPackage` object causes the owning side of the `one-to-one` association to be overwritten with a proxy reference, which means that `DDC2363Order#$services_package` now references the old lazy-loaded proxy.

This happens because the query to load the `DDC2363ServicesPackage` implicitly fetch-joins `DDC2363Order`, and hydration causes the existing `DDC2363Order#$services_package` to be replaced.

Very very very tricky, and I don't think this is fixable without completely disabling:

 * lazy-loading of cascade-removed proxies
 * fetch-joins on one-to-one associations

**Comment by Ocramius** [comment]
Targeting for `3.0`, since this likely needs a decision on whether to apply a BC break and either:

 * stop lazy-loading entries that are marked for cascade removal
 * stop implicitly fetch-joining one-to-one associations when the inverse side is selected

**Comment by Ocramius** [comment]
Rebased/simplified branch at https://github.com/doctrine/doctrine2/tree/DDC-2363_cascade-removal-triggering-one-to-one-fetch-join-during-flush

**Comment by Ocramius** [comment]
Also checked on `develop`: same bug there, so far.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    public function setServicesPackage(DDC2363ServicesPackage $services_package)
+    {
+        $this->services_package = $services_package;
```
Comment (carnage) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: In most cases, I'd set the order property on the existing record to null here to ensure that the object graph in memory is correct. This would sidestep the issue above as the lazy loading wouldn't be triggered in the cascade remove operation. It also means that any code with a reference to the old services package object can't use it (assuming you put some guards in place to throw exceptions if the order property is null)

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: That's a workaround though, not a fix

Comment (carnage) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: I'd argue that you are ensuring correct system behaviour. If you weren't persisting to the DB; you should set the order property null on the old object as it is no longer linked to one. Otherwise you may end up with weird edge cases if something else in your system maintains a reference to the old object. (Which is actually what we are seeing here from a different angle)

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: Ah, I see what you mean, so the lazy-loading no longer occurs in ORM-internals but rather when the inverse side is set to `null` by the code here.

That could work, yes.

I'm wondering if the test scenario in this particular ticket has a valid use-case scenario then.

Also note that this will break if you have logic like following (valid, IMO):

```php
    public function setServicesPackage(DDC2363ServicesPackage $services_package)
    {
        $originalServicesPackage = $this->services_package;
        $this->services_package = $services_package;
        $services_package->order = $this;
        $originalServicesPackage->order = null; // lazy-loading will reset the property
    }
```

Comment (mmenozzi) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: Hi @Ocramius, I don't remember the whole story about this issue but I remember that we had this issue in production on a real (big) project so I think it's a real use case. Why do you wonder if it's a valid use-case? I mean, with correct orphan removal configuration I'd expect to have those orphan removed.

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: It's a question of references to be cleaned. This code is wrong:

```php
$this->services_package = $services_package;
```

This code is right:

```php
if ($this->services_package) {
    $this->services_package->order = null;
}

$this->services_package = $services_package;
```

This doesn't deny the ORM bug, it just denies this specific scenario (the one in the test case).

Comment (mmenozzi) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: @Ocramius so even with the right code there is a bug?

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: Yes, but I need a test scenario that doesn't have your userland bug in it.

In my opinion, the bug is that fetches are being performed during a `flush()` (that's already super-weird, but understandable)

Comment (carnage) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: I'd suggest that the issue may lie in lazy loading overwriting a property which has been changed. Is it possible to fix that somehow?

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: We could disable fetc-joining during flush operations, but I need to check if that's even configurable.

Comment (mmenozzi) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: @Ocramius so do you need that I push the right code on my fork for this PR?

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: No need - the issue is clear, I provided a newer/rebased PR for it, but the action to be taken is still unclear.

Comment (mmenozzi) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php]: Ok.
